### PR TITLE
feat: Support split_kwargs from DataOp for advanced cross-validation

### LINKED ISSUE #2884: Use the cv splitter and scoring of skrub learners
it is possible in skrub DataOps to specify a cross-validation splitter and a scoring metric (and any additional arguments they need):

```python
import skore
import skrub
from sklearn.dummy import DummyClassifier
from sklearn.model_selection import LeaveOneGroupOut

df = skrub.datasets.toy_products()
data = skrub.var("df", df)
groups = data["seller"]
# Configure the cv splitter
X = data[["description", "price"]].skb.mark_as_X(
    cv=LeaveOneGroupOut(), split_kwargs={"groups": groups}
)
sample_weight = X["price"]
y = data["category"].skb.mark_as_y()
pred = (
    X.skb.apply(DummyClassifier(), y=y)
    # Configure scoring
    .skb.with_scoring("accuracy")
    .skb.with_scoring(
        "accuracy",
        kwargs={"sample_weight": sample_weight},
        name="weighted_accuracy",
    )
)

cv = pred.skb.cross_validate() # splits grouping by seller, scores with and without sample weights
print(cv)
```

```
   fit_time  score_time  test_accuracy  test_weighted_accuracy
0  0.001563    0.003420       0.666667                0.594595
1  0.001237    0.002599       0.666667                0.909091
```

At the moment skore ignores this configuration (it overrides it with its own defaults).

```python
report = skore.evaluate(pred.skb.make_learner(), data={"df": df})
summary = report.metrics.summarize().frame()
print(summary)
```

```
                             SkrubLearner
                                     mean
Metric           Label
Accuracy                         0.500000
Precision        electronics     0.500000
                 tools           0.000000
Recall           electronics     1.000000
                 tools           0.000000
ROC AUC                          0.500000
Log loss                         0.836988
Brier score                      0.312500
Fit time (s)                     0.002137
Predict time (s)                 0.000943
```

It would be helpful if it could use the configured splitter by default when it exists, and either use the configured scorers or add them to the default ones in .metrics.summarize() (e.g. weighted_accuracy would show up in the summary above, and splitting would use the groups)

### PR DESCRIPTION
**Pull request by glemaitre**
closes #2884 

Support more advanced cross-validation by leveraging `split_kwargs` from skrub `DataOp`.

### CODE REVIEW

File: examples/technical_details/plot_skrub_data_op_cv.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#
+# We use the toy products dataset and group products by seller. The goal is to
+# assess generalization to new sellers with
+# :class:`~sklearn.model_selection.LeaveOneGroupOut`.
+import skrub
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import LeaveOneGroupOut
+
+df = skrub.datasets.toy_products()
+data = skrub.var("df", df)
```
Comment (jeromedockes) [inline examples/technical_details/plot_skrub_data_op_cv.py]: if we don't show it we don't really need the df here
```suggestion
data = skrub.var("df")
```

File: skore/src/skore/_sklearn/_cross_validation/report.py [hunk 2]
```diff
@@ -21,19 +21,28 @@
...
+    PositiveLabel,
+    SKLearnCrossValidator,
+    _DefaultType,
+)
 from skore._utils._fixes import _validate_joblib_parallel_params
 from skore._utils._parallel import delayed
 from skore._utils._progress_bar import track
-from skore._utils._skrub import eval_X_y, is_skrub_learner, to_estimator, to_learner
+from skore._utils._skrub import (
+    eval_X_y,
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_cross_validation/report.py]: i think instead of eval_X_y and then getting the indices separately we should use `.skb.find_X_y` which will return X, y, the splitter and split_kwargs and evaluate all in one go

File: skore/src/skore/_sklearn/_cross_validation/report.py [hunk 3]
```diff
@@ -243,8 +252,26 @@ def __init__(
             _check_estimator_and_data(clone(estimator), X, y, data)
         )
         self._pos_label = pos_label
-        self._splitter = check_cv(splitter, y, classifier=is_classifier(estimator))
-        self._split_indices = tuple(self._splitter.split(self.X, self.y))
+        if self._initialized_with_data_op:
+            cv_for_indices = splitter if splitter is not None else _DEFAULT
+            self._split_indices = resolve_data_op_split_indices(
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_cross_validation/report.py]: we could get that at the same time as the data

File: skore/src/skore/_sklearn/_cross_validation/report.py [hunk 4]
```diff
@@ -225,7 +234,7 @@ def __init__(
         y: ArrayLike | None = None,
         data: dict | None = None,
         pos_label: PositiveLabel | None = None,
-        splitter: int | SKLearnCrossValidator | Generator | None = None,
+        splitter: int | SKLearnCrossValidator | Generator | _DefaultType | None = None,
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_cross_validation/report.py]: i'm not sure why we need to accept DefaultTYpe here instead of just None?

File: skore/src/skore/_sklearn/_cross_validation/report.py [hunk 5]
```diff
@@ -659,11 +686,11 @@ def input_data(self) -> dict:
         return self._data.copy()
 
     @property
-    def splitter(self) -> SKLearnCrossValidator:
+    def splitter(self) -> SKLearnCrossValidator | None:
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_cross_validation/report.py]: if we use find_X_y here we can return the splitter that was produced by evaluating the result

File: skore/src/skore/_sklearn/_cross_validation/report.py [hunk 6]
```diff
@@ -739,7 +766,11 @@ def to_markdown(self) -> str:
                 "splitter_repr": (
                     repr(self.splitter)
                     if self.splitter is not None
-                    else f"{len(self.split_indices)} folds"
+                    else (
+                        "from DataOp"
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/_cross_validation/report.py]: not important but we could still show the number of splits

File: skore/src/skore/_sklearn/evaluate.py [hunk 7]
```diff
@@ -64,7 +71,11 @@ def evaluate(
         (e.g. ``{"X": X_df, "other_table": df, ...}``).
 
     splitter : float, int, str, or cross-validation object, default=0.2
-        Determines how the data is split:
+        Determines how the data is split. When omitted, a skrub learner whose
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/evaluate.py:75]: the 80/20 is really restricting a shufflesplit to its first split. i wonder if we could have an api to do that for all splitters, ie separate the "do only the first split" part from the choice of splitter

File: skore/src/skore/_sklearn/evaluate.py [hunk 8]
```diff
@@ -203,6 +214,20 @@ def evaluate(
...
+    if splitter is _DEFAULT:
+        data_op = get_data_op(estimator)
+        if data_op is not None and data_op_has_explicit_cv(data_op):
+            return CrossValidationReport(
+                estimator,
+                X,
+                y,
+                data=data,
+                pos_label=pos_label,
+                splitter=_DEFAULT,
```
Comment (jeromedockes) [inline skore/src/skore/_sklearn/evaluate.py]: I guess we could pass splitter=None here and not need to handle _DEFAULT in the crossvalidation report?

File: skore/src/skore/_sklearn/evaluate.py (hunk 7 above)
Comment (jeromedockes) [reply skore/src/skore/_sklearn/evaluate.py:75]: also I find the behavior of this default a bit hard to understand 🤔

**Review by jeromedockes — DISMISSED** [review]
it's a great improvement thanks @glemaitre  . i had a few comments but I'm also happy to merge and open a follow-up pr myself

File: skore/src/skore/_sklearn/evaluate.py (hunk 8 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/evaluate.py]: Since we do a single 20% train test split by default, I think we might be in trouble for the others not?

File: skore/src/skore/_sklearn/_cross_validation/report.py (hunk 2 above)
Comment (jeromedockes) [reply skore/src/skore/_sklearn/_cross_validation/report.py]: something like

```python
X_y_cv = skrub.as_data_op(pred.skb.find_X_y()).skb.eval(env)
data = {"_skrub_X": X_y_cv["X"], "_skrub_y": X_y_cv["y"]}
if X_y_cv.get("cv") is not None:
    split_indices = list(X_y_cv["cv"].split(X_y_cv["X"], X_y_cv["y"], **X_y_cv["split_kwargs"]))
    splitter = X_y_cv["cv"]
```

File: skore/src/skore/_sklearn/evaluate.py (hunk 8 above)
Comment (jeromedockes) [reply skore/src/skore/_sklearn/evaluate.py]: i just have the impression there is some manipulation around None and default that could be simplified, here we pass default, which crossvalidationreport converts to default and passes to resolve_data_split_indices which converts back to None 🤔 i guess the only place we really need a mechanism to distinguish None vs "not provided" is here in evaluate because of the 0.2 default

File: skore/src/skore/_sklearn/evaluate.py (hunk 7 above)
Comment (jeromedockes) [reply skore/src/skore/_sklearn/evaluate.py:75]: we will discuss that in another issue

**Review by jeromedockes — DISMISSED** [review]

**Review by jeromedockes — APPROVED** [review]
